### PR TITLE
Improve performance for grids with permission checks

### PR DESCRIPTION
**Pull request by sreichel**
### Description (*)

Some grids (eg orders) check whether user is allowed to click/view ... but its valididated for each row.

With this PR its only checked once.

### CODE REVIEW

**Comment by kiatng** [comment]
How about implementing it at the session level:
https://github.com/OpenMage/magento-lts/blob/0ac3f19a23eed02fbb27cc5586254a5eb7d30af9/app/code/core/Mage/Admin/Model/Session.php#L224-L250

The improvement would not be limited to grid.

In this PR, line 229 `if ($user && $acl) { ` is bypassed, but I suspect this check is unnecessary. If `$user` is empty, it won't get pass `Mage_Adminhtml_Controller_Action::predispatch()`.

**Comment by sreichel** [comment]
@kiatng i did some tests with xdebug and cant see any acl resource is checked multiple times. Need some advice.

**Comment by sreichel** [comment]
Sample Data: Sales Order Grid with 46 orders

Before PR:

<table border="1" cellpadding="2" cellspacing="1" width="90%" rules="rows" bordercolor="#bdc7d8" align="center">
<tbody><tr bgcolor="#bdc7d8" align="right"><th align="left"><nobr><a href="/xhprof/index.php?run=65bd80e8636d7&amp;symbol=Mage_Adminhtml_Block_Sales_Order_Grid%3A%3AgetRowUrl&amp;source=ddev&amp;sort=fn">Function Name</a></nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=65bd80e8636d7&amp;symbol=Mage_Adminhtml_Block_Sales_Order_Grid%3A%3AgetRowUrl&amp;source=ddev&amp;sort=ct">Calls</a></nobr></th><th class="vwbar"><nobr>Calls%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=65bd80e8636d7&amp;symbol=Mage_Adminhtml_Block_Sales_Order_Grid%3A%3AgetRowUrl&amp;source=ddev&amp;sort=wt">Incl. Wall Time<br>(microsec)</a></nobr></th><th class="vwbar"><nobr>IWall%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=65bd80e8636d7&amp;symbol=Mage_Adminhtml_Block_Sales_Order_Grid%3A%3AgetRowUrl&amp;source=ddev&amp;sort=cpu">Incl. CPU<br>(microsecs)</a></nobr></th><th class="vwbar"><nobr>ICpu%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=65bd80e8636d7&amp;symbol=Mage_Adminhtml_Block_Sales_Order_Grid%3A%3AgetRowUrl&amp;source=ddev&amp;sort=mu">Incl.<br>MemUse<br>(bytes)</a></nobr></th><th class="vwbar"><nobr>IMemUse%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=65bd80e8636d7&amp;symbol=Mage_Adminhtml_Block_Sales_Order_Grid%3A%3AgetRowUrl&amp;source=ddev&amp;sort=pmu">Incl.<br> PeakMemUse<br>(bytes)</a></nobr></th><th class="vwbar"><nobr>IPeakMemUse%</nobr></th></tr><tr bgcolor="#e0e0ff"><td><b><i><center>Current Function</center></i></b></td></tr><tr><td><a href="">Mage_Adminhtml_Block_Sales_Order_Grid::getRowUrl</a></td><td class="vbar">46</td>
<td class="vbar">17.9%</td>
<td class="vbbar">135,638</td>
<td class="vbbar">11.7%</td>
<td class="vbar">106,614</td>
<td class="vbar">9.4%</td>
<td class="vbar">10,536</td>
<td class="vbar">0.2%</td>
<td class="vbar">5,880</td>
<td class="vbar">0.1%</td>
</tr><tr bgcolor="#ffffff"><td style="text-align:right;color:blue">Exclusive Metrics  for Current Function</td><td class="vbar"></td><td class="vbar"></td><td type="Child" metric="wt" class="vbbar">807</td>
<td type="Child" metric="wt" class="vbbar">0.6%</td>
<td type="Child" metric="cpu" class="vbar">0</td>
<td type="Child" metric="cpu" class="vbar">0.0%</td>
<td type="Child" metric="mu" class="vbar">-4,296</td>
<td type="Child" metric="mu" class="vbar">-40.8%</td>
<td type="Child" metric="pmu" class="vbar">64</td>
<td type="Child" metric="pmu" class="vbar">1.1%</td>
</tr><tr bgcolor="#e0e0ff"><td><b><i><center>Parent function</center></i></b></td></tr><tr><td><a href="/xhprof/index.php?run=65bd80e8636d7&amp;symbol=Mage_Core_Block_Template%3A%3AfetchView%402&amp;source=ddev">Mage_Core_Block_Template::fetchView@2</a></td><td type="Parent" metric="ct" class="vbar">46</td>
<td type="Parent" metric="ct" class="vbar">100.0%</td>
<td type="Parent" metric="wt" class="vbbar">135,638</td>
<td type="Parent" metric="wt" class="vbbar">100.0%</td>
<td type="Parent" metric="cpu" class="vbar">106,614</td>
<td type="Parent" metric="cpu" class="vbar">100.0%</td>
<td type="Parent" metric="mu" class="vbar">10,536</td>
<td type="Parent" metric="mu" class="vbar">100.0%</td>
<td type="Parent" metric="pmu" class="vbar">5,880</td>
<td type="Parent" metric="pmu" class="vbar">100.0%</td>
</tr><tr bgcolor="#e0e0ff"><td><b><i><center>Child functions</center></i></b></td></tr><tr><td><a href="/xhprof/index.php?run=65bd80e8636d7&amp;symbol=Mage_Core_Block_Abstract%3A%3AgetUrl&amp;source=ddev">Mage_Core_Block_Abstract::getUrl</a></td><td type="Child" metric="ct" class="vbar">46</td>
<td type="Child" metric="ct" class="vbar">25.0%</td>
<td type="Child" metric="wt" class="vbbar">87,147</td>
<td type="Child" metric="wt" class="vbbar">64.2%</td>
<td type="Child" metric="cpu" class="vbar">72,718</td>
<td type="Child" metric="cpu" class="vbar">68.2%</td>
<td type="Child" metric="mu" class="vbar">7,944</td>
<td type="Child" metric="mu" class="vbar">75.4%</td>
<td type="Child" metric="pmu" class="vbar">4,328</td>
<td type="Child" metric="pmu" class="vbar">73.6%</td>
</tr><tr bgcolor="#e5e5e5"><td><a href="/xhprof/index.php?run=65bd80e8636d7&amp;symbol=Mage_Admin_Model_Session%3A%3AisAllowed&amp;source=ddev">Mage_Admin_Model_Session::isAllowed</a></td><td type="Child" metric="ct" class="vbar">46</td>
<td type="Child" metric="ct" class="vbar">25.0%</td>
<td type="Child" metric="wt" class="vbbar">46,690</td>
<td type="Child" metric="wt" class="vbbar">34.4%</td>
<td type="Child" metric="cpu" class="vbar">33,896</td>
<td type="Child" metric="cpu" class="vbar">31.8%</td>
<td type="Child" metric="mu" class="vbar">3,160</td>
<td type="Child" metric="mu" class="vbar">30.0%</td>
<td type="Child" metric="pmu" class="vbar">1,488</td>
<td type="Child" metric="pmu" class="vbar">25.3%</td>
</tr><tr><td><a href="/xhprof/index.php?run=65bd80e8636d7&amp;symbol=Mage_Core_Model_Abstract%3A%3AgetId&amp;source=ddev">Mage_Core_Model_Abstract::getId</a></td><td type="Child" metric="ct" class="vbar">46</td>
<td type="Child" metric="ct" class="vbar">25.0%</td>
<td type="Child" metric="wt" class="vbbar">872</td>
<td type="Child" metric="wt" class="vbbar">0.6%</td>
<td type="Child" metric="cpu" class="vbar">0</td>
<td type="Child" metric="cpu" class="vbar">0.0%</td>
<td type="Child" metric="mu" class="vbar">584</td>
<td type="Child" metric="mu" class="vbar">5.5%</td>
<td type="Child" metric="pmu" class="vbar">0</td>
<td type="Child" metric="pmu" class="vbar">0.0%</td>
</tr><tr bgcolor="#e5e5e5"><td><a href="/xhprof/index.php?run=65bd80e8636d7&amp;symbol=Mage%3A%3AgetSingleton&amp;source=ddev">Mage::getSingleton</a></td><td type="Child" metric="ct" class="vbar">46</td>
<td type="Child" metric="ct" class="vbar">25.0%</td>
<td type="Child" metric="wt" class="vbbar">122</td>
<td type="Child" metric="wt" class="vbbar">0.1%</td>
<td type="Child" metric="cpu" class="vbar">0</td>
<td type="Child" metric="cpu" class="vbar">0.0%</td>
<td type="Child" metric="mu" class="vbar">3,144</td>
<td type="Child" metric="mu" class="vbar">29.8%</td>
<td type="Child" metric="pmu" class="vbar">0</td>
<td type="Child" metric="pmu" class="vbar">0.0%</td>
</tr></tbody></table>

After PR:

<table border="1" cellpadding="2" cellspacing="1" width="90%" rules="rows" bordercolor="#bdc7d8" align="center">
<tbody><tr bgcolor="#bdc7d8" align="right"><th align="left"><nobr><a href="/xhprof/index.php?run=65bd80fce708f&amp;symbol=Mage_Adminhtml_Block_Sales_Order_Grid%3A%3AgetRowUrl&amp;source=ddev&amp;sort=fn">Function Name</a></nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=65bd80fce708f&amp;symbol=Mage_Adminhtml_Block_Sales_Order_Grid%3A%3AgetRowUrl&amp;source=ddev&amp;sort=ct">Calls</a></nobr></th><th class="vwbar"><nobr>Calls%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=65bd80fce708f&amp;symbol=Mage_Adminhtml_Block_Sales_Order_Grid%3A%3AgetRowUrl&amp;source=ddev&amp;sort=wt">Incl. Wall Time<br>(microsec)</a></nobr></th><th class="vwbar"><nobr>IWall%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=65bd80fce708f&amp;symbol=Mage_Adminhtml_Block_Sales_Order_Grid%3A%3AgetRowUrl&amp;source=ddev&amp;sort=cpu">Incl. CPU<br>(microsecs)</a></nobr></th><th class="vwbar"><nobr>ICpu%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=65bd80fce708f&amp;symbol=Mage_Adminhtml_Block_Sales_Order_Grid%3A%3AgetRowUrl&amp;source=ddev&amp;sort=mu">Incl.<br>MemUse<br>(bytes)</a></nobr></th><th class="vwbar"><nobr>IMemUse%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=65bd80fce708f&amp;symbol=Mage_Adminhtml_Block_Sales_Order_Grid%3A%3AgetRowUrl&amp;source=ddev&amp;sort=pmu">Incl.<br> PeakMemUse<br>(bytes)</a></nobr></th><th class="vwbar"><nobr>IPeakMemUse%</nobr></th></tr><tr bgcolor="#e0e0ff"><td><b><i><center>Current Function</center></i></b></td></tr><tr><td><a href="">Mage_Adminhtml_Block_Sales_Order_Grid::getRowUrl</a></td><td class="vbar">46</td>
<td class="vbar">21.8%</td>
<td class="vbbar">94,850</td>
<td class="vbbar">8.6%</td>
<td class="vbar">106,904</td>
<td class="vbar">9.8%</td>
<td class="vbar">12,240</td>
<td class="vbar">0.3%</td>
<td class="vbar">7,584</td>
<td class="vbar">0.2%</td>
</tr><tr bgcolor="#ffffff"><td style="text-align:right;color:blue">Exclusive Metrics  for Current Function</td><td class="vbar"></td><td class="vbar"></td><td type="Child" metric="wt" class="vbbar">649</td>
<td type="Child" metric="wt" class="vbbar">0.7%</td>
<td type="Child" metric="cpu" class="vbar">0</td>
<td type="Child" metric="cpu" class="vbar">0.0%</td>
<td type="Child" metric="mu" class="vbar">792</td>
<td type="Child" metric="mu" class="vbar">6.5%</td>
<td type="Child" metric="pmu" class="vbar">368</td>
<td type="Child" metric="pmu" class="vbar">4.9%</td>
</tr><tr bgcolor="#e0e0ff"><td><b><i><center>Parent function</center></i></b></td></tr><tr><td><a href="/xhprof/index.php?run=65bd80fce708f&amp;symbol=Mage_Core_Block_Template%3A%3AfetchView%402&amp;source=ddev">Mage_Core_Block_Template::fetchView@2</a></td><td type="Parent" metric="ct" class="vbar">46</td>
<td type="Parent" metric="ct" class="vbar">100.0%</td>
<td type="Parent" metric="wt" class="vbbar">94,850</td>
<td type="Parent" metric="wt" class="vbbar">100.0%</td>
<td type="Parent" metric="cpu" class="vbar">106,904</td>
<td type="Parent" metric="cpu" class="vbar">100.0%</td>
<td type="Parent" metric="mu" class="vbar">12,240</td>
<td type="Parent" metric="mu" class="vbar">100.0%</td>
<td type="Parent" metric="pmu" class="vbar">7,584</td>
<td type="Parent" metric="pmu" class="vbar">100.0%</td>
</tr><tr bgcolor="#e0e0ff"><td><b><i><center>Child functions</center></i></b></td></tr><tr><td><a href="/xhprof/index.php?run=65bd80fce708f&amp;symbol=Mage_Core_Block_Abstract%3A%3AgetUrl&amp;source=ddev">Mage_Core_Block_Abstract::getUrl</a></td><td type="Child" metric="ct" class="vbar">46</td>
<td type="Child" metric="ct" class="vbar">33.3%</td>
<td type="Child" metric="wt" class="vbbar">91,712</td>
<td type="Child" metric="wt" class="vbbar">96.7%</td>
<td type="Child" metric="cpu" class="vbar">102,916</td>
<td type="Child" metric="cpu" class="vbar">96.3%</td>
<td type="Child" metric="mu" class="vbar">7,944</td>
<td type="Child" metric="mu" class="vbar">64.9%</td>
<td type="Child" metric="pmu" class="vbar">7,104</td>
<td type="Child" metric="pmu" class="vbar">93.7%</td>
</tr><tr bgcolor="#e5e5e5"><td><a href="/xhprof/index.php?run=65bd80fce708f&amp;symbol=Mage_Adminhtml_Block_Widget_Grid%3A%3AcanView&amp;source=ddev">Mage_Adminhtml_Block_Widget_Grid::canView</a></td><td type="Child" metric="ct" class="vbar">46</td>
<td type="Child" metric="ct" class="vbar">33.3%</td>
<td type="Child" metric="wt" class="vbbar">1,473</td>
<td type="Child" metric="wt" class="vbbar">1.6%</td>
<td type="Child" metric="cpu" class="vbar">3,988</td>
<td type="Child" metric="cpu" class="vbar">3.7%</td>
<td type="Child" metric="mu" class="vbar">2,920</td>
<td type="Child" metric="mu" class="vbar">23.9%</td>
<td type="Child" metric="pmu" class="vbar">80</td>
<td type="Child" metric="pmu" class="vbar">1.1%</td>
</tr><tr><td><a href="/xhprof/index.php?run=65bd80fce708f&amp;symbol=Mage_Core_Model_Abstract%3A%3AgetId&amp;source=ddev">Mage_Core_Model_Abstract::getId</a></td><td type="Child" metric="ct" class="vbar">46</td>
<td type="Child" metric="ct" class="vbar">33.3%</td>
<td type="Child" metric="wt" class="vbbar">1,016</td>
<td type="Child" metric="wt" class="vbbar">1.1%</td>
<td type="Child" metric="cpu" class="vbar">0</td>
<td type="Child" metric="cpu" class="vbar">0.0%</td>
<td type="Child" metric="mu" class="vbar">584</td>
<td type="Child" metric="mu" class="vbar">4.8%</td>
<td type="Child" metric="pmu" class="vbar">32</td>
<td type="Child" metric="pmu" class="vbar">0.4%</td>
</tr></tbody></table>

File: app/code/core/Mage/Adminhtml/Block/Widget/Grid.php [hunk 1]
```diff
@@ -220,6 +220,13 @@ class Mage_Adminhtml_Block_Widget_Grid extends Mage_Adminhtml_Block_Widget
         ],
     ];
 
+    /**
+     * Allow edit status
+     *
+     * @var array<string, bool>|null
+     */
+    protected ?array $isAllowed = null;
```
Comment (fballiano) [inline app/code/core/Mage/Adminhtml/Block/Widget/Grid.php:228]:
```suggestion
    protected array $isAllowed = [];
```

wouldn't it be simpler to always have an array? the same way as https://github.com/OpenMage/magento-lts/pull/3785?

File: app/code/core/Mage/Adminhtml/Block/Widget/Grid.php [hunk 2]
```diff
@@ -220,6 +220,13 @@ class Mage_Adminhtml_Block_Widget_Grid extends Mage_Adminhtml_Block_Widget
         ],
     ];
 
+    /**
+     * Allow edit status
```
Comment (fballiano) [inline app/code/core/Mage/Adminhtml/Block/Widget/Grid.php:224]:
```suggestion
     * This array caches the status of the isAllow() method, for every acl path
```

or something like that

File: app/code/core/Mage/Adminhtml/Block/Widget/Grid.php [hunk 3]
```diff
@@ -1886,4 +1893,29 @@ public function getLimitOptions(): array
...
+     */
+    private function _isAllowed(string $aclPath): bool
+    {
+        if ($this->isAllowed === null || !isset($this->isAllowed[$aclPath])) {
+            /** @var Mage_Admin_Model_Session $session */
+            $session = Mage::getSingleton('admin/session');
+            $this->isAllowed[$aclPath] = $session->isAllowed($aclPath);
+        }
+
+        return $this->isAllowed[$aclPath];
```
Comment (fballiano) [inline app/code/core/Mage/Adminhtml/Block/Widget/Grid.php:1919]: if the array was always an array

```suggestion
        if (!isset($this->isAllowed[$aclPath]) {
            /** @var Mage_Admin_Model_Session $session */
            $session = Mage::getSingleton('admin/session');
            $this->isAllowed[$aclPath] = $session->isAllowed($aclPath);
        }

        return $this->isAllowed[$aclPath];
```
?

**Comment by sreichel** [comment]
Okay ... working from smartphone was not ideal ... 😂 Fix it later.

**Pull request closed without merging**